File: watcher/models/event.go
package models

import "time"

// EventType represents the kind of file system change.
type EventType string

const (
	EventCreate EventType = "CREATE"
	EventModify EventType = "MODIFY"
	EventDelete EventType = "DELETE"
	EventRename EventType = "RENAME"
)

// FileEvent represents a single file change.
type FileEvent struct {
	Type        EventType `json:"type"`
	Path        string    `json:"path"`
	OldPath     string    `json:"old_path,omitempty"`
	Project     string    `json:"project"`
	MachineID   string    `json:"machine_id"`
	Timestamp   int64     `json:"timestamp"`
	ContentHash string    `json:"content_hash,omitempty"`
	Content     string    `json:"content,omitempty"`  // optional: file content for small files
	Entities    []Entity  `json:"entities,omitempty"` // AST extraction results
}

// Entity represents an extracted code symbol.
type Entity struct {
	Name      string `json:"name"`
	Type      string `json:"type"` // function, class, struct, interface, import
	Line      int    `json:"line"`
	StartLine int    `json:"start_line,omitempty"`
	EndLine   int    `json:"end_line,omitempty"`
}

// BatchPayload is the HTTP request body sent to Hub.
type BatchPayload struct {
	MachineID string      `json:"machine_id"`
	Project   string      `json:"project"`
	Events    []FileEvent `json:"events"`
	SentAt    time.Time   `json:"sent_at"`
}

// TombstoneEvent is a special delete event.
type TombstoneEvent struct {
	MachineID string `json:"machine_id"`
	Project   string `json:"project"`
	Path      string `json:"path"`
}

File: hub/mcp_server/models/event.py
"""Event models shared between Watcher and Hub API."""

from pydantic import BaseModel


class Entity(BaseModel):
    name: str
    type: str
    line: int
    start_line: int | None = None
    end_line: int | None = None


class Relation(BaseModel):
    type: str
    source: str | None = None
    target: str
    target_type: str | None = None
    line: int | None = None
    confidence: str | None = None
    layer: str | None = None
    status: str | None = None
    symbol_id: str | None = None
    target_ref: str | None = None
    package: str | None = None
    language: str | None = None


class FileEvent(BaseModel):
    type: str
    path: str
    old_path: str | None = None
    project: str
    machine_id: str
    timestamp: int
    content_hash: str | None = None
    content: str | None = None
    entities: list[Entity] | None = None
    relations: list[Relation] | None = None


class BatchPayload(BaseModel):
    machine_id: str
    project: str
    events: list[FileEvent]
    sent_at: str | None = None
